libsql-server/src/namespace/schema_lock.rs

```diff
@@ -0,0 +1,95 @@
+use std::sync::Arc;
+
+use hashbrown::{hash_map::Entry, HashMap};
+use parking_lot::Mutex;
+
+use super::NamespaceName;
+
+#[derive(Default)]
+pub struct SchemaLocksRegistry {
+    locks: Arc<Mutex<HashMap<NamespaceName, Arc<tokio::sync::RwLock<()>>>>>,
+}
+
+enum SchemaLockKind {
+    Shared(tokio::sync::OwnedRwLockReadGuard<()>),
+    Exclusive(tokio::sync::OwnedRwLockWriteGuard<()>),
+}
+
+pub struct SchemaLock {
+    schema: NamespaceName,
+    _guard: Option<SchemaLockKind>,
+    locks: Arc<Mutex<HashMap<NamespaceName, Arc<tokio::sync::RwLock<()>>>>>,
+}
+
+impl Drop for SchemaLock {
+    fn drop(&mut self) {
+        let mut locks = self.locks.lock();
+        match locks.entry(self.schema.clone()) {
+            Entry::Occupied(entry) => {
+                // there's only two ref left: the maps, and ours
+                if Arc::strong_count(entry.get()) == 2 {
```

Comment:
If we cannot make this safe by compiler type checks, at least we should write comments in both the places pointing to this line, so it won't be forgotten during refactoring.